File: python/stripe_agent_toolkit/openai/toolkit.py
"""Stripe Agent Toolkit."""

from typing import List, Optional
from pydantic import PrivateAttr
import json

from agents import FunctionTool


from ..api import StripeAPI
from ..tools import tools
from ..configuration import Configuration, is_tool_allowed
from .tool import StripeTool
from .hooks import BillingHooks

class StripeAgentToolkit:
    _tools: List[FunctionTool] = PrivateAttr(default=[])
    _stripe_api: StripeAPI = PrivateAttr(default=None)

    def __init__(
        self, secret_key: str, configuration: Optional[Configuration] = None
    ):
        super().__init__()

        context = configuration.get("context") if configuration else None

        self._stripe_api = StripeAPI(secret_key=secret_key, context=context)

        filtered_tools = [
            tool for tool in tools if is_tool_allowed(tool, configuration)
        ]

        self._tools = [
            StripeTool(self._stripe_api, tool)
            for tool in filtered_tools
        ]

    def get_tools(self) -> List[FunctionTool]:
        """Get the tools in the toolkit."""
        return self._tools

    def billing_hook(self, type: Optional[str] = None, customer: Optional[str] = None, meter: Optional[str] = None, meters: Optional[dict[str, str]] = None) -> BillingHooks:
        return BillingHooks(self._stripe_api, type, customer, meter, meters)

File: typescript/src/ai-sdk/toolkit.ts
import StripeAPI from '../shared/api';
import tools from '../shared/tools';
import {isToolAllowed, type Configuration} from '../shared/configuration';
import type {
  CoreTool,
  LanguageModelV1StreamPart,
  Experimental_LanguageModelV1Middleware as LanguageModelV1Middleware,
} from 'ai';
import StripeTool from './tool';

type StripeMiddlewareConfig = {
  billing?: {
    type?: 'token';
    customer: string;
    meters: {
      input?: string;
      output?: string;
    };
  };
};

class StripeAgentToolkit {
  private _stripe: StripeAPI;

  tools: {[key: string]: CoreTool};

  constructor({
    secretKey,
    configuration,
  }: {
    secretKey: string;
    configuration: Configuration;
  }) {
    this._stripe = new StripeAPI(secretKey, configuration.context);
    this.tools = {};

    const context = configuration.context || {};
    const filteredTools = tools(context).filter((tool) =>
      isToolAllowed(tool, configuration)
    );

    filteredTools.forEach((tool) => {
      // @ts-ignore
      this.tools[tool.method] = StripeTool(
        this._stripe,
        tool.method,
        tool.description,
        tool.parameters
      );
    });
  }

  middleware(config: StripeMiddlewareConfig): LanguageModelV1Middleware {
    const bill = async ({
      promptTokens,
      completionTokens,
    }: {
      promptTokens: number;
      completionTokens: number;
    }) => {
      if (config.billing) {
        if (config.billing.meters.input) {
          await this._stripe.createMeterEvent({
            event: config.billing.meters.input,
            customer: config.billing.customer,
            value: promptTokens.toString(),
          });
        }
        if (config.billing.meters.output) {
          await this._stripe.createMeterEvent({
            event: config.billing.meters.output,
            customer: config.billing.customer,
            value: completionTokens.toString(),
          });
        }
      }
    };

    return {
      wrapGenerate: async ({doGenerate}) => {
        const result = await doGenerate();

        if (config.billing) {
          await bill(result.usage);
        }

        return result;
      },

      wrapStream: async ({doStream}) => {
        const {stream, ...rest} = await doStream();

        const transformStream = new TransformStream<
          LanguageModelV1StreamPart,
          LanguageModelV1StreamPart
        >({
          async transform(chunk, controller) {
            if (chunk.type === 'finish') {
              if (config.billing) {
                await bill(chunk.usage);
              }
            }

            controller.enqueue(chunk);
          },
        });

        return {
          stream: stream.pipeThrough(transformStream),
          ...rest,
        };
      },
    };
  }

  getTools(): {[key: string]: CoreTool} {
    return this.tools;
  }
}

export default StripeAgentToolkit;
